### PR TITLE
planner: stop pushing Agg down through Projection if substitution fail

### LINKED ISSUE #50926: planner: the optimizer returns 'cannot find column' error when aggregating view with json columns
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

<!-- a step by step guide for reproducing the bug. -->

```
mysql> create table t (a int);
Query OK, 0 rows affected (0.03 sec)

mysql> create or replace algorithm=undefined view v (a,b) AS
    ->   select 1 as a, json_object('k', '0') as b from t;
Query OK, 0 rows affected (0.05 sec)

mysql> select sum(json_extract(b, '$.path')) from v group by a;
ERROR 1105 (HY000): Can't find column Column#4 in schema Column: [test.t._tidb_rowid] Unique key: []
```

### 2. What did you expect to see? (Required)
No error

### 3. What did you see instead (Required)
ERROR 1105 (HY000): Can't find column Column#4 in schema Column: [test.t._tidb_rowid] Unique key: []

### 4. What is your TiDB version? (Required)

<!-- Paste the output of SELECT tidb_version() -->
```
+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| tidb_version()                                                                                                                                                                                                                                                                 |
+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Release Version: v6.5.6
Edition: Community
Git Commit Hash: 8445821f105477d5e31de9d8451b7fbf35a02cd1
Git Branch: HEAD
UTC Build Time: 2024-02-02 04:13:50
GoVersion: go1.21.0
Race Enabled: false
TiKV Min Version: 6.2.0-alpha
Check Table Before Drop: false
Store: unistore |
+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
```

### PR DESCRIPTION
**Pull request by qw4990**
### What problem does this PR solve?

Issue Number: close #50926

Problem Summary: planner: stop pushing Agg down through Projection if substitution fail

### What changed and how does it work?

planner: stop pushing Agg down through Projection if substitution fail

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: pkg/planner/core/rule_aggregation_push_down.go [hunk 1]
```diff
@@ -558,7 +558,11 @@ func (a *aggregationPushDownSolver) aggPushDown(p LogicalPlan, opt *logicalOptim
 				noSideEffects := true
 				newGbyItems := make([]expression.Expression, 0, len(agg.GroupByItems))
 				for _, gbyItem := range agg.GroupByItems {
-					newGbyItems = append(newGbyItems, expression.ColumnSubstitute(ctx, gbyItem, proj.schema, proj.Exprs))
+					_, failed, groupBy := expression.ColumnSubstituteImpl(ctx, gbyItem, proj.schema, proj.Exprs, true)
+					if failed {
+						noSideEffects = false
```
Comment (AilinKid) [inline pkg/planner/core/rule_aggregation_push_down.go:563]: adding break will be a quick out

Comment (qw4990) [reply pkg/planner/core/rule_aggregation_push_down.go:563]: done, thx!

**Review by AilinKid — APPROVED** [review]

**Review by hawkingrei — APPROVED** [review]

**Review by guo-shaoge — APPROVED** [review]

**Review by XuHuaiyu — APPROVED** [review]

**Pull request merged**